src/Common/src/Interop/OSX/Interop.libproc.cs

```diff
@@ -0,0 +1,542 @@
+        {
+            internal proc_bsdinfo    pbsd;
+            internal proc_taskinfo   ptinfo;
+        }
+
+        // From proc_info.h
+        [StructLayout(LayoutKind.Sequential)]
+        internal unsafe struct proc_threadinfo
+        {
+            internal ulong      pth_user_time;
+            internal ulong      pth_system_time;
+            internal int        pth_cpu_usage;
+            internal int        pth_policy;
+            internal int        pth_run_state;
+            internal int        pth_flags;
+            internal int        pth_sleep_time;
+            internal int        pth_curpri;
+            internal int        pth_priority;
+            internal int        pth_maxpriority;
+            internal fixed byte pth_name[MAXTHREADNAMESIZE];
+        }
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct proc_fdinfo
+        {
+            internal int proc_fd;
+            internal uint proc_fdtype;
+        }
+
+        /// <summary>
+        /// Queries the OS for the PIDs for all running processes
+        /// </summary>
+        /// <param name="buffer">A pointer to the memory block where the PID array will start</param>
+        /// <param name="buffersize">The length of the block of memory allocated for the PID array</param>
+        /// <returns>Returns the number of elements (PIDs) in the buffer</returns>
+        [DllImport(Interop.Libraries.libproc, SetLastError = true)]
+        private static unsafe extern int proc_listallpids(
+            int*    pBuffer, 
+            int     buffersize);
+
+        /// <summary>
+        /// Queries the OS for the list of all running processes and returns the PID for each
+        /// </summary>
+        /// <returns>Returns a list of PIDs corresponding to all running processes</returns>
+        internal static unsafe int[] proc_listallpids()
+        {
+            // Get the number of processes currently running to know how much data to allocate
+            int numProcesses = proc_listallpids(null, 0);
+            if (numProcesses <= 0)
+            {
+                throw new System.ComponentModel.Win32Exception(Marshal.GetLastWin32Error(), SR.CantGetAllPids);
+            }
+
+            int[] processes = null;
+
+            do
+            {
+                // Create a new array for the processes (plus a 10% buffer in case new processes have spawned)
+                processes = new int[(int)(numProcesses + (numProcesses * .10))];
+                fixed (int* pBuffer = processes)
+                {
+                    numProcesses = proc_listallpids(pBuffer, processes.Length * Marshal.SizeOf<int>());
+                    if (numProcesses <= 0)
+                    {
+                        throw new System.ComponentModel.Win32Exception(Marshal.GetLastWin32Error(), SR.CantGetAllPids);
+                    }
+                }
+            }
+            while (numProcesses == processes.Length);
```

Comment:
I seem to remember you previously having a comment about why you were doing this, that you needed to be able to differentiate the case where the # of processes == processes.Length vs # of processes > processes.Length, both of which would result in numProcesses == process.Length?  I thought that was a helpful comment; consider putting it back.